File: java/src/main/java/com/epi/MatrixSearchTest.java
package com.epi;
// @pg_import
import com.epi.utils.AbstractTestOptions;
import com.epi.utils.AbstractTestStream;
import com.epi.utils.JsonTestOptions;
import com.epi.utils.TestType;

import java.util.Arrays;
import java.util.List;
// @pg_end
// @pg_ignore:1
import static com.epi.MatrixSearch.matrixSearch;

public class MatrixSearchTest {
  // @pg_ignore
  // @pg_include:MatrixSearch.java
  // @pg_end
  private static void unitTest(AbstractTestOptions options, String description,
                               List<List<Integer>> input, int x, boolean expected) {
    AbstractTestStream stream = options.getStream();
    stream.startTest(TestType.NORMAL, description);
    stream.getInputWriter().
        writePair("matrix", input).
        writePair("x", x).
        close();
    stream.registerExpectedOutput(expected);

    try {
      boolean result = matrixSearch(input, x);
      stream.registerUserOutput(result, expected == result);
    } catch (NullPointerException e) {
      stream.registerNullPointerException();
    } catch (Exception e) {
      stream.registerUnhandledException();
    }

    stream.endTest();
  }

  public static void directedTests(AbstractTestOptions options) {
    options.startTests(0, "Matrix search");

    unitTest(options, "Trivial test #1", Arrays.asList(Arrays.asList(1)), 1, true);
    unitTest(options, "Trivial test #1", Arrays.asList(Arrays.asList(1)), 2, false);
    unitTest(options, "1D matrix test #1", Arrays.asList(Arrays.asList(1, 3, 5, 7, 9, 9, 11)), 11, true);
    unitTest(options, "1D matrix test #2", Arrays.asList(Arrays.asList(1, 3, 5, 7, 9, 9, 11)), 9, true);
    unitTest(options, "1D matrix test #3", Arrays.asList(Arrays.asList(1, 3, 5, 7, 9, 9, 11)), 1, true);
    unitTest(options, "1D matrix test #4", Arrays.asList(Arrays.asList(1, 3, 5, 7, 9, 9, 11)), 8, false);
    unitTest(options, "1D matrix test #5", Arrays.asList(
        Arrays.asList(1),
        Arrays.asList(3),
        Arrays.asList(5),
        Arrays.asList(7),
        Arrays.asList(9),
        Arrays.asList(9),
        Arrays.asList(11)), 11, true);
    unitTest(options, "1D matrix test #6", Arrays.asList(
        Arrays.asList(1),
        Arrays.asList(3),
        Arrays.asList(5),
        Arrays.asList(7),
        Arrays.asList(9),
        Arrays.asList(9),
        Arrays.asList(11)), 9, true);
    unitTest(options, "1D matrix test #7", Arrays.asList(
        Arrays.asList(1),
        Arrays.asList(3),
        Arrays.asList(5),
        Arrays.asList(7),
        Arrays.asList(9),
        Arrays.asList(9),
        Arrays.asList(11)), 1, true);
    unitTest(options, "1D matrix test #8", Arrays.asList(
        Arrays.asList(1),
        Arrays.asList(3),
        Arrays.asList(5),
        Arrays.asList(7),
        Arrays.asList(9),
        Arrays.asList(9),
        Arrays.asList(11)), 8, false);
    unitTest(options, "Full test #1", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 1, true);
    unitTest(options, "Full test #2", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 5, true);
    unitTest(options, "Full test #3", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 3, true);
    unitTest(options, "Full test #4", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 10, true);
    unitTest(options, "Full test #5", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 1100, true);
    unitTest(options, "Full test #6", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 3000, true);
    unitTest(options, "Full test #7", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), Integer.MAX_VALUE,
        true);
    unitTest(options, "Full test #8", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 50, true);
    unitTest(options, "Full test #9", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 30, true);
    unitTest(options, "Full test #10", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 21, true);
    unitTest(options, "Full test #11", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 13, false);
    unitTest(options, "Full test #12", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), 2500, false);
    unitTest(options, "Full test #13", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)),
        Integer.MAX_VALUE - 1, false);
    unitTest(options, "Full test #14", Arrays.asList(
        Arrays.asList(1, 2, 3, 4, 5),
        Arrays.asList(2, 3, 4, 5, 6),
        Arrays.asList(10, 20, 30, 40, 50),
        Arrays.asList(15, 21, 30, 40, 5000),
        Arrays.asList(1100, 2000, 3000, 4000, Integer.MAX_VALUE)), Integer.MIN_VALUE,
        false);
    options.endTests();
  }

  public static void main(String[] args) {
    directedTests(new JsonTestOptions(System.out));
  }

}

File: cpptest/Matrix_searchTest.cc
// Copyright (c) 2016 Elements of Programming Interviews. All rights reserved.
#include "test_toolkit/test-options.h"
// @pg_ignore:1
#include "Matrix_search.cc"
// @pg_include:Matrix_search.cc

void UnitTest(TestSentry::Ptr& sentry, const char* description,
              const vector<vector<int>>& input, int x, bool expected) {
  TestStream::Ptr stream = sentry->GetTestStream(TestType::NORMAL, description);
  stream->GetInputWriter()->
      WritePair("matrix", input).
      WritePair("x", x);
  stream->RegisterExpectedOutput(expected);
  try {
    bool result = MatrixSearch(input, x);
    stream->RegisterUserOutput(result, result == expected);
  }
  catch (...) {
    stream->RegisterUnhandledException();
  }
}

void DirectedTests(const TestOptions& options) {
  TestSentry::Ptr sentry = options.GetTestSentry(0, "Matrix search");

  UnitTest(sentry, "Trivial test #1", {{1}}, 1, true);
  UnitTest(sentry, "Trivial test #1", {{1}}, 2, false);
  UnitTest(sentry, "1D matrix test #1", {{1, 3, 5, 7, 9, 9, 11}}, 11, true);
  UnitTest(sentry, "1D matrix test #2", {{1, 3, 5, 7, 9, 9, 11}}, 9, true);
  UnitTest(sentry, "1D matrix test #3", {{1, 3, 5, 7, 9, 9, 11}}, 1, true);
  UnitTest(sentry, "1D matrix test #4", {{1, 3, 5, 7, 9, 9, 11}}, 8, false);
  UnitTest(sentry, "1D matrix test #5", {{1},
                                         {3},
                                         {5},
                                         {7},
                                         {9},
                                         {9},
                                         {11}}, 11, true);
  UnitTest(sentry, "1D matrix test #6", {{1},
                                         {3},
                                         {5},
                                         {7},
                                         {9},
                                         {9},
                                         {11}}, 9, true);
  UnitTest(sentry, "1D matrix test #7", {{1},
                                         {3},
                                         {5},
                                         {7},
                                         {9},
                                         {9},
                                         {11}}, 1, true);
  UnitTest(sentry, "1D matrix test #8", {{1},
                                         {3},
                                         {5},
                                         {7},
                                         {9},
                                         {9},
                                         {11}}, 8, false);
  UnitTest(sentry, "Full test #1", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 1, true);
  UnitTest(sentry, "Full test #2", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 5, true);
  UnitTest(sentry, "Full test #3", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 3, true);
  UnitTest(sentry, "Full test #4", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 10, true);
  UnitTest(sentry, "Full test #5", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 1100, true);
  UnitTest(sentry, "Full test #6", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 3000, true);
  UnitTest(sentry, "Full test #7", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, numeric_limits<int>::max(),
           true);
  UnitTest(sentry, "Full test #8", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 50, true);
  UnitTest(sentry, "Full test #9", {{1,    2,    3,    4,    5},
                                    {2,    3,    4,    5,    6},
                                    {10,   20,   30,   40,   50},
                                    {15,   21,   30,   40,   5000},
                                    {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 30, true);
  UnitTest(sentry, "Full test #10", {{1,    2,    3,    4,    5},
                                     {2,    3,    4,    5,    6},
                                     {10,   20,   30,   40,   50},
                                     {15,   21,   30,   40,   5000},
                                     {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 21, true);
  UnitTest(sentry, "Full test #11", {{1,    2,    3,    4,    5},
                                     {2,    3,    4,    5,    6},
                                     {10,   20,   30,   40,   50},
                                     {15,   21,   30,   40,   5000},
                                     {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 13, false);
  UnitTest(sentry, "Full test #12", {{1,    2,    3,    4,    5},
                                     {2,    3,    4,    5,    6},
                                     {10,   20,   30,   40,   50},
                                     {15,   21,   30,   40,   5000},
                                     {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, 2500, false);
  UnitTest(sentry, "Full test #13", {{1,    2,    3,    4,    5},
                                     {2,    3,    4,    5,    6},
                                     {10,   20,   30,   40,   50},
                                     {15,   21,   30,   40,   5000},
                                     {1100, 2000, 3000, 4000, numeric_limits<int>::max()}},
           numeric_limits<int>::max() - 1, false);
  UnitTest(sentry, "Full test #14", {{1,    2,    3,    4,    5},
                                     {2,    3,    4,    5,    6},
                                     {10,   20,   30,   40,   50},
                                     {15,   21,   30,   40,   5000},
                                     {1100, 2000, 3000, 4000, numeric_limits<int>::max()}}, numeric_limits<int>::min(),
           false);
}

int main(int argc, char* argv[]) {
  DirectedTests(TestOptions(&cout));
  return 0;
}
